drivers/pppou.c

```diff
@@ -0,0 +1,627 @@
+/*
+ * Phoenix-RTOS --- networking stack
+ *
+ * PPP over Serial (null modem) driver
+ *
+ * Copyright 2018, 2021 Phoenix Systems
+ * Author: Marek Białowąs, Gerard Świderski
+ *
+ * %LICENSE%
+ */
+
+#include "netif-driver.h"
+
+#include <netif/ppp/pppapi.h>
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <string.h>
+#include <unistd.h>
+#include <strings.h>
+#include <sys/threads.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+#include <termios.h>
+#include <syslog.h>
+
+/*
+ *  Setup full-duplex connection between host and device with RS232, uart USB-TTL adapter, direct TTL-TTL, bluetooth-UART/TTL, etc.
+ *
+ *  1. On device (phoenix-rtos) side, addr=10.0.0.2, use:
+ *     lwip pppou:/dev/uart3:<speed>:up
+ *     (speed defaults to 115200 if ommited, 'up' brings interface up after setup, also optional)
+ *
+ *  2. On host (linux, bsd) side, addr=10.0.0.1, use e.g.:
+ *     pppd /dev/ttyUSB0 <speed> 10.0.0.1:10.0.0.2 lock local nodetach noauth debug dump nocrtscts nodefaultroute maxfail 0 holdoff 1
+ *
+ *  Replace <speed> with any of the baud rate supported by the system, e.g. 9600, 115200, 230400 or 460800.
+ */
+
+
+enum conn_state_e {
+	CONN_STATE_DISCONNECTING,
+	CONN_STATE_DISCONNECTED,
+	CONN_STATE_CONNECTING,
+	CONN_STATE_CONNECTED,
+};
+
+enum cfg_flag_e {
+	CFG_FLAG_DEFAULT_UP = 0x01
+};
+
+typedef struct
+{
+	struct netif *netif;
+	ppp_pcb* ppp;
+
+	const char *serial_dev;
+	speed_t serial_speed;
+	int fd;
+
+	volatile int thread_running;
+	volatile int want_connected;
+	volatile int conn_state;
+	handle_t lock, cond;
+
+	unsigned char main_loop_stack[4096] __attribute__((aligned(8)));
+} pppou_priv_t;
+
+
+#if 0
+
+#define log_debug(fmt, ...) syslog(LOG_DEBUG, fmt, ##__VA_ARGS__)
+#define log_at(fmt, ...)    syslog(LOG_INFO, fmt, ##__VA_ARGS__)
+#define log_info(fmt, ...)  syslog(LOG_INFO, fmt, ##__VA_ARGS__)
+#define log_warn(fmt, ...)  syslog(LOG_WARNING, fmt, ##__VA_ARGS__)
+#define log_error(fmt, ...) syslog(LOG_ERR, fmt, ##__VA_ARGS__)
+
+#else
+
+#define log_debug(fmt, ...)
+#define log_at(fmt, ...)
+#define log_info(fmt, ...)
+#define log_warn(fmt, ...)
+#define log_error(fmt, ...)
+
+#endif
+
+#define PPPOU_READ_DATA_TIMEOUT_STEP_MS  10
+
+#define PPPOU_TRYOPEN_SERIALDEV_SEC      3
+#define PPPOU_CONNECT_RETRY_SEC          5
+#define PPPOU_CONNECT_CMD_RETRY_MS       500
+
+
+/****** serial handling ******/
+
+speed_t serial_speed_from_string(const char str[])
```

Comment:
```suggestion
speed_t serial_speed_from_string(const char *str)
```
I think it will be better.

Reply:
This way `char str[]` is denoted as C style string parameter _(array of characters `\0` terminated)_ , what You suggest is a pointer to the very first signed character of a memory - nearly the same, but not exactly. See also, I'm consistent with `serial_open()`, where `devname` is also a string there. String names are decayed to pointers, hence, You can use pointers to manipulate elements of the string. The difference is when it comes to initializers, then of course assigning `char str[]` a pointer is invalid.
As absolute classic:  `int main(int argc, char *argv[], char *envp[])` of course one can write this as `main(int argc, char **argv, char **envp)`, nearly the same, but ... I'm tied to the classic.